### PR TITLE
rust implementation (repl)

### PR DESCRIPTION
**Pull request by matheusfillipe**
This pr aims to provide a minimal hanb implementation in rust. This includes:

- created rust project
    - Hanb Binary
    - Hanb rust Libary
- Simple cli to print boards
- Basic hanb repl

Please play around with the repl `cargo run` inside the rust folder and provide feedback if this fullfills the hanb requirements.

I am leaving these extra features as todos:

* Add repl history
* Add repl completion
* Use board from cli or stdin or file (save and load)

### CODE REVIEW

**Comment by handyc** [comment]
This is a really nice start to a Rust implementation of hanb and I would like to merge it with the main project soon.
A few changes that would improve it in my opinion:
- [x] given an input of less than 64 characters, editing in the REPL should allow for any cell to be edited (currently appears to allow only editing of cells that are created with the initial string, so if I input e.g. a 42-character string, it does not allow me to edit cell 56 etc.)
- [x] resolving a cell to the next lower level should always return a new board, even if that cell is not already defined with a return mapping
- [x] still to be determined: if 'a' is the smallest possible thing (a quantum foam particle), then what should be the return value for a query on an 'a'? (my initial feeling was to return a random value to simulate quantum randomness but in conversations with matheusfillipe we did not reach an agreement on this idea [yet])
- [x] any values on a hanb should be allowable and resolve to legal/valid hanb boards, so for example on an O-size (city) board, a value of '.' (universe) should be interpreted as something that makes sense within that particular context, e.g. a new random value of the appropriate size, or the smallest possible unit 'a', or something else (still trying to figure out the best solution here)
- [x] input characters that are not part of the hanb character set of a-z, A-Z, 0-9, - and . should resolve to sensible allowed characters, e.g. "This ham sandwich is incredible!" should resolve to "This-ham-sandwich-is-incredible." i.e. any punctuation mark aside from the hyphen should resolve to a period (full stop) and any space should resolve to a hyphen. In this way natural language inputs can be used as seeds for hanbs.
- [ ] a hanb board that is defined with object positions and object types together is known as a 'situation' (a static arrangement of items). The transition from one situation to a different situation is known as an 'event'. This feature is the basis for all movement in hanb.
- [x] the board display should properly reflect the hexagonal nature of hanb (it is not sufficient to display all of the cells, they should also be properly spaced to be perceived easily as a hexagonal arrangement of cells)
- [x] it would be nice to be able to submit a query to hanb and receive a value back without entering a REPL, allowing hanb queries to be submitted through bash scripts etc.

**Comment by matheusfillipe** [comment]
- [ ] 1
- [ ] 2
- [ ] 3

File: rust/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,302 @@
+use std::cmp::min;
+
+const BOARD_SIZE: usize = 61;
+const DIAGONAL: u8 = 9;
```
Comment (matheusfillipe) [inline rust/src/lib.rs]:
```suggestion
const DIAMETER: u8 = 9;
```

**Review by matheusfillipe — COMMENTED** [review]
This is just me testing a plugin

**Comment by matheusfillipe** [comment]
More plugin testing

File: rust/src/lib.rs (hunk 1 above)
Comment (matheusfillipe) [reply rust/src/lib.rs:4]: What is this

**Comment by matheusfillipe** [comment]
No way man

**Comment by handyc** [comment]
re: current implementation of saving and loading:
it is nice to have an indication of which level the board was on when it was saved, but it should also be possible to load a board of one level as another level (this is a major feature of hanb since the beginning).
For example, if I begin a board on level L as: AaaaAAAAaAAAAAAAAAAAAAKKJJJJJKKJJHHJJaaaaaaaaaaa
I should be able to load it later as level O. In this way relationships between objects can be used across different scales.

**Comment by matheusfillipe** [comment]
On that case should we truncate bigger values then the board you got in @handyc? What if tree depth of the importing relationship goes beyond the level you are trying to import it from. Seems to me this causes to many complicated issues. Better focus on just importing and exporting the entire universes first.

**Comment by matheusfillipe** [comment]
@handyc I believe this is feature complete now. Please check for bugs and other small improvements and review the code.

**Comment by handyc** [comment]
> @handyc I believe this is feature complete now. Please check for bugs and other small improvements and review the code.


below is my previous comment from when I was using an earlier version, here are my new comments having updated
properly to the newest version:

- cells should always resolve to a new board when going (d)own -- currently undefined boards simply don't resolve
- undefined boards should resolve to the smallest unit 'a'
- going (u)p should always be allowed, even if the higher board is not defined
- going up to a higher undefined board should take into account the 3 values at the end of the board string (61,62,63)


-------------
Current issues:

choosing the start level and then setting the board appears to reset the start level to '.' = universe size
e.g. set the starting level to O (city) and then use 'b' to set a board, the container size appears to reset to '.' in all cases

previously there was an issue in which setting the container size to O and then initializing a board with O size objects was allowed -- not sure if this bug is also still lurking behind the other one,
proper behavior should be to reduce objects that are too large to the largest possible object for a given container,
e.g. if an O size object is set within an O size board, the object should be reduced to N size

setting a board with a string that is too short (e.g. 30 characters instead of 64 characters) should cause the remaining
characters to be filled out with 'a' (quantum foam) and not '.'

a new container size for '.' should be include as '#'

queries on 'a' particles should not return a board but instead return a new random 'a' type (this can be added
later when types are introduced)

not able to test other things like saving/loading properly at this time due to the issue with level reset

**Comment by handyc** [comment]
I am still in the process of testing, but I think we are rapidly nearing the point where it would be suitable to merge this version with the main repo. It really looks nice!

**Comment by matheusfillipe** [comment]
Great. The initial level setting was reset and defaulted to the highest possible level. This is a simplification and avoids problems related to going to upper levels and with the board initialization. 
I've also addressed the glitch with the '.' being set.
The value of the color of the current board is returned when you reach the bottom or upper level.
I am leaving the implementation fo the random and types for a future iteration.

For now I believe we can go with making sure the sizes are consistent and the load/save/import/export are ok.

Also leave some feedback for changing the names of commands, correcting documentation (the readme), the help menus, or do those changes and we can merge them into this commit.

**Comment by handyc** [comment]
adding these items to the PR comments just so that we don't lose track of them:

[x] Do not truncate color as sizes
[x] # is not a color (color codes are 3-character codes from the hanb 64 character alphabet only)
[x] When a board is resized loop over all cells and inner cells recursively resizing their sizes by the same offset and if smaller than a delete them and their children
[x] "save | s  [filename: string = hanb.hsit] -> Saves the current situation to a file" -> "save current explored map to a file"
(a situation is the state of object positions and objects types within one hanb board only)

**Comment by matheusfillipe** [comment]
@handyc I believe the last bugs were solved now. I've opted to also resize the inner cells not only when a cell is scalled down but also when its size increases, everything inside should go up by the same offset. Please take some time for testing it. I believe this is getting close to good enough. Try to find more bugs. 

Remember that you can test the scaling bug with:

```sh
echo "b A\nd\nb zzz\nu\nb\ns" | cargo run -- -s
```

Thank you and good luck.

**Comment by handyc** [comment]
> @handyc I believe the last bugs were solved now. I've opted to also resize the inner cells not only when a cell is scalled down but also when its size increases, everything inside should go up by the same offset. Please take some time for testing it. I believe this is getting close to good enough. Try to find more bugs.
> 
> Remember that you can test the scaling bug with:
> 
> ```shell
> echo "b A\nd\nb zzz\nu\nb\ns" | cargo run -- -s
> ```
> 
> Thank you and good luck.

Great, currently testing, will post an update soon. Thank you!

**Comment by matheusfillipe** [comment]
Added workflow for checking, testing and deploying to docker. You will have to add your own dockerhub user and token to github secrets to be able to have the docker image automatically deployed on push to the rust folders. Also the bot on irc is up to date and ready for you to test.

**Comment by handyc** [comment]
This comment is just a placeholder to say that I am still planning to merge this pull request, still need to test all functions thoroughly. Sorry for the long delay!

**Pull request merged**